pkg/yurthub/kubernetes/meta/restmapper.go

```diff
@@ -0,0 +1,259 @@
+/*
+Copyright 2021 The OpenYurt Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package meta
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/openyurtio/openyurt/pkg/yurthub/storage"
+	"github.com/openyurtio/openyurt/pkg/yurthub/storage/disk"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/klog"
+)
+
+const (
+	CacheDynamicRESTMapperKey = "_internal/restmapper/cache-crd-restmapper.conf"
+	SepForGVR                 = "/"
+)
+
+var (
+	// unsafeSchemeRESTMapper is used to store the mapping relationship between GVK and GVR in scheme
+	// It is not updatable and is only used for the judgment of scheme resources
+	unsafeSchemeRESTMapper = NewDefaultRESTMapperFromScheme()
+	ErrGVRNotRecognized    = errors.New("GroupVersionResource is not recognized")
+)
+
+// RESTMapperManager is responsible for managing different kind of RESTMapper
+type RESTMapperManager struct {
+	sync.RWMutex
+	storage storage.Store
+	// UnsafeDefaultRESTMapper is used to save the GVK and GVR mapping relationships of built-in resources
+	unsafeDefaultRESTMapper *meta.DefaultRESTMapper
+	// dynamicRESTMapper is used to save the GVK and GVR mapping relationships of Custom Resources
+	dynamicRESTMapper map[schema.GroupVersionResource]schema.GroupVersionKind
+}
+
+func NewDefaultRESTMapperFromScheme() *meta.DefaultRESTMapper {
+	s := scheme.Scheme
+	defaultGroupVersions := s.PrioritizedVersionsAllGroups()
+	mapper := meta.NewDefaultRESTMapper(defaultGroupVersions)
+	// enumerate all supported versions, get the kinds, and register with the mapper how to address
+	// our resources.
+	for _, gv := range defaultGroupVersions {
+		for kind := range s.KnownTypes(gv) {
+			// Only need to process non-list resources
+			if !strings.HasSuffix(kind, "List") {
+				// Since RESTMapper is only used for mapping GVR to GVK information,
+				// the scope field is not involved in actual use,
+				// so all scope are currently set to meta.RESTScopeNamespace
+				scope := meta.RESTScopeNamespace
+				mapper.Add(gv.WithKind(kind), scope)
+			}
+		}
+	}
+	return mapper
+}
+
+func NewRESTMapperManager(rootDir string) *RESTMapperManager {
+	var dm map[schema.GroupVersionResource]schema.GroupVersionKind
+	dStorage, err := disk.NewDiskStorage(rootDir)
```

Comment:
why not make `disk storage` as parameter for `NewRESTMapperManager`?

Reply:
The code is updated, the `disk storage` is passed into the `NewRESTMapperManager` as a variable.